### PR TITLE
Vesting

### PR DESCRIPTION
**Pull request by riley-stride**
## Summary
This PR runs through the vesting flow using the [core cosmos-sdk vesting module](https://docs.cosmos.network/main/modules/auth/05_vesting.html). We do not alter the vesting module code in this PR, only discuss which vesting module and vesting accounts to use and show instructions on how to kick off vesting. 

## Details
- added a csv file containing addresses to vest to and amounts of tokens to vest, see `scripts-local/vesting/vesting_accts.csv`
- added a script that kicks off `DelayedVesting` to these accounts, with `CLIFF` and `DURATION` paramaterized. See `scripts-local/vesting/add_vesting_accts.sh`
- Within `scripts-local/init_stride.sh`, run the vesting script to start vesting

## Testing
To vest tokens linearly with a cliff to a few accounts, run the standard local init process with `make init-local build=s`. 
You should see this output in the shell
```
VESTING ACCOUNTS:
Vesting  100000000ustrd to stride1ze3jp6ut8ce9qrxtthh9r5cy632mn9rqchp8qy, with a 00D:02M:00.00S cliff, then linearly over 00D:05M:00.00S.
Vesting  50000000000ustrd to stride1qj0jws6fycdhxcja27tfwazdp26pq8mwpyt829, with a 00D:02M:00.00S cliff, then linearly over 00D:05M:00.00S.
Vesting  350000000ustrd to stride1qcrp9lefnnmtgx60nrm5quc6x88xp0u9ha7wp2, with a 00D:02M:00.00S cliff, then linearly over 00D:05M:00.00S.
```
The check the vesting accounts on the block explorer e.g. http://localhost:8080/stride/account/stride1ze3jp6ut8ce9qrxtthh9r5cy632mn9rqchp8qy. 

See `x/vesting/README.md` for detailed testing instructions. Copied below:

_Note that the amount vested is taken out of the total amount genesis'd to the account, it is NOT additional to the amount genesis'd (in this case, that means stride1ze3jp6ut8ce9qrxtthh9r5cy632mn9rqchp8qy will have `100000000ustrd` total tokens, al of which will be vesting)_

A simple way to check how much has vested is to look at the account on the block explorer (with our local ping.pub, setup from https://github.com/Stride-Labs/explorer, that's http://localhost:8080/stride/account/stride1ze3jp6ut8ce9qrxtthh9r5cy632mn9rqchp8qy). 

Another way is to attempt to bank send the full acct balance to another addr before the vest has completed. Inspecting that failed tx will show you how much is available to send (which tells you how much has vested thus far). To test vesting via the bank send approach, swap out an address in the vesting csv for one you control, then during vesting

```
strided-local='build/strided --home ./scripts-local/state/stride'
strided-local tx bank send val1 stride1ft20pydau82pgesyl9huhhux307s9h3078692y <TOTAL_BALANCE>ustrd --chain-id STRIDE --keyring-backend test -y
strided-local q tx <TX_HASH_FROM_PREV_TX>
```
Once vesting completes 7 min from genesis, bank sending the full account balance should succeed!

## Open questions
- Should we switch to using Osmosis' [version of the vesting module](https://github.com/osmosis-labs/cosmos-sdk/tree/osmosis-main/x/auth/vesting
- https://docs.evmos.org/modules/vesting/ )? Need to look through it to see what (if anything) they added
- Do we want/need clawbacks?

## For mainnet

We want delayed vesting for all vesting accounts. We simply need to, for each acct 
1. calc the amount to vest
2. have the recipient set up an addr on testnet, save their seed phrase and send us the address
3. set up vesting for that account with the following params 
    ```
    VESTING_START_TIME=1685592000 # June 1st 2023
    VESTING_END_TIME=1780286400 # June 1st 2026
    VESTING_AMT='XXXustrd'
    VESTER_ADDR='YYY'
    $STRIDE_CMD add-genesis-account $VESTER_ADDR $VESTING_AMT --vesting-start-time $VESTING_START_TIME --vesting-end-time $VESTING_END_TIME --vesting-amount $VESTING_AMT
    ```

### CODE REVIEW

File: scripts-local/vesting/add_vesting_accts.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+convertsecs() {
+ h=$(bc <<< "${1}/3600")
```
Comment (vish-stride) [inline scripts-local/vesting/add_vesting_accts.sh:4]: this is fancy, I didn't know the <<< operator existed!

**Review by vish-stride — APPROVED** [review]
this looks great to me! very thorough README as well

only question from my end, do we need to add all the vesting accounts at genesis, or is it possible to add any after chain is launched? totally fine if it's genesis, just means we should be more meticulous about it before launch

**Comment by asalzmann** [comment]
Sorry for the delay - will review this tomorrow!

**Comment by asalzmann** [comment]
I think the params might be off
```
VESTING_START_TIME=1685592000 # June 1st 2023
VESTING_END_TIME=1780286400 # June 1st 2026
VESTING_AMT='XXXustrd'
VESTER_ADDR='YYY'
$STRIDE_CMD add-genesis-account $VESTER_ADDR $VESTING_AMT --vesting-start-time $VESTING_START_TIME --vesting-end-time $VESTING_END_TIME --vesting-amount $VESTING_AMT
```

otherwise this lgtm!

**Review by asalzmann — APPROVED** [review]

**Comment by riley-stride** [comment]
> this looks great to me! very thorough README as well
> 
> only question from my end, do we need to add all the vesting accounts at genesis, or is it possible to add any after chain is launched? totally fine if it's genesis, just means we should be more meticulous about it before launch

Genesis afaiu!

**Pull request merged**